### ISSUE TITLE
RandomBut attribute

### ISSUE THREAD

**Issue by rileyyy** [nunit/nunit#4102]
I would like some way of using the Random attribute but providing it an exception.  This could be either an optional parameter or a new [RandomBut(...)] attribute. A single value would be great, but it would be even niftier if the attribute accepted a collection. I.e.
```cs
// I'd love an optional parameter
public void someTest( [Random(-1, 1, 3, butNot: 0)] int parameter) { ... }

// or a collection
public void someTest( [Random(-1, 1, 3, new List<int> { 0 })] int parameter) { ... }

// or even a params argument, if that isn't too confusing
public void someTest( [Random(-1, 1, 3, 0, 0.1, -0.1 )] double parameter) { ... }
```

Contrived example would be testing the following, where I would expect to have a test case specifically to `Assert.Throws<DivideByZeroException>()`.  
```cs
double Reciprocal(double denominator)
{
  if (denominator is 0)
    throw new DivideByException("Can't divide by zero doofus.")
  return 1 / denominator;
}
```

Although, it is very unlikely that the Randomizer would present exactly 0, I hope the idea of the example is clear enough.  I suppose another example could involve an enum with the exception of a certain value.  I understand that its possible to  use 
```cs 
if (denominator is 0) 
  Assert.Inconclusive()
```
But, in a large enough TestFixture it feels repetitive to need to continue rewriting Assert.Inconclusive() clauses.

**Comment by stevenaw** [comment]
I agree using `Assert.Inconclusive` is not ideal here, especially if that's a value which doesn't apply in a given scenario, but I'm wondering if there's a built-in way here to help. @rileyyy Have you considered using multiple tests to cover the different ranges on either side of where you want to exclude, or perhaps using a `ValueSource`?
```
double Reciprocal_Negatives( [Random(-10, -1, 3)] double denominator) {}
double Reciprocal_Positives( [Random(1, 10, 3)] double denominator) {}
```

```
double Reciprocal( [ValueSource("MyFunction")] double denominator) {}

public static IEnumerable MyFunction() {
  for (var i =0; i < 3; i++) {
     double value = 0.0;

     while (value != 0.0) {
         value = TestContext.CurrentContext.Randomizer.NextDouble(-10, 10);
     }
 
    yield return value;
  }
}
```

I haven't tested the code to catch compiler issues, but in both cases it might do what you are hoping for

**Comment by rileyyy** [comment]
I have considered multiple tests,  unfortunately my given example was a bit easier than my current use case, and it would ~4x my test count which I think would have made it a bit too unwieldy to maintain.  I wasn't actually aware of the ValueSource attribute though, which seems like a good fit for what I need.
It seems in that case that my suggestion is not much more than a small convince for readability.  I am happy to close in that case, if this change doesn't provide additional benefit.

**Comment by jnm2** [comment]
What's the viability of a separate `[ExcludeValues(...)]` attribute? Maybe the framework could then delete test cases generated by any other parameter attribute.

**Comment by stevenaw** [comment]
That's an interesting idea.

A couple things off the top of my head we'd have to consider:
1. For things like `RandomAttribute`, would we keep trying until we generate a valid value?
2. For user-defined values like `TestCaseSource` would we just ignore the value if it is returned? Would it be "Inconclusive"?
3. Would it apply to class-level and method-level params equally (`TestFixtureSource` vs `TestCaseSource`)
4. Would parameter type conversion rules apply (`"2022-04-22"` vs `new DateTime(2022, 4, 22)`)
